### PR TITLE
[JENKINS-69549] Margins for headers and paragraphs make descriptions …

### PR DESCRIPTION
**Pull request by frankie139506**
See [JENKINS-69549](https://issues.jenkins.io/browse/JENKINS-69549).
Before:
image

After:
image

Another example:
image

### Proposed changelog entries

- Reduce spacing in help files

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [X] (If applicable) Jira issue is well described
- [X] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change) and are in the imperative mood. [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  - Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [X] Appropriate autotests or explanation to why this change has no tests
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadoc, as appropriate.
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")` if applicable.
- [ ] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content-Security-Policy directives (see [documentation on jenkins.io](https://www.jenkins.io/doc/developer/security/csp/)).
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`:

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are accurate, human-readable, and in the imperative mood
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).

<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/7078"></a>

### CODE REVIEW

**Comment by daniel-beck** [comment]
With changes like this it would be interesting to know why the original change was done in the first place. Requesting @janfaracik as a reviewer to point out what this change in turn will break ;-)

**Comment by janfaracik** [comment]
> With changes like this it would be interesting to know why the original change was done in the first place. Requesting @janfaracik as a reviewer to point out what this change in turn will break ;-)

The original change was made to make page headers have a margin consistent with the page margin. This causes issues like the above however. 

For https://issues.jenkins.io/browse/JENKINS-69549 there's two issues:
- the titles are using `h1` - this should be changed to be correct for the heading level
- the margin is set incorrectly (as mentioned above)

(I haven't tried this PR yet) but I believe a better fix would be to change the value of `--section-padding` in the help component class, e.g. half the current value.

**Comment by frankie139506** [comment]
> * the titles are using `h1` - this should be changed to be correct for the heading level
I have changed to h3 for the title.
> * the margin is set incorrectly (as mentioned above)
I have made some change and I think this change make more reasonable.

**Comment by daniel-beck** [comment]
> the titles are using `h1` - this should be changed to be correct for the heading level

This is impractical to implement across the ecosystem.

**Comment by janfaracik** [comment]
> > the titles are using `h1` - this should be changed to be correct for the heading level
> 
> This is impractical to implement across the ecosystem. We need to expect that all levels of `h*`, `p`, `div`, and basic formatting are used freely in help files, and deal with it.

Would it be possible (and relatively straightforward) to dynamically changing the header levels when clicking the❓button?

**Comment by daniel-beck** [comment]
> Would it be possible (and relatively straightforward) to dynamically change the header levels when clicking the❓button?

We already rewrite the responses some in https://github.com/jenkinsci/jenkins/blob/04b46d1c68513e4a105b5acdfb7534c89da94eea/war/src/main/webapp/scripts/hudson-behavior.js#L990-L1005 so it should be feasible. Is this preferable to e.g. having styles conditional on `h*` being inside/output the help block, and if so, why?

File: war/src/main/less/base/style.less [hunk 1]
```diff
@@ -573,7 +573,7 @@ pre.console {
   }
 
   p:last-of-type {
-    margin-bottom: 0;
+    margin-bottom: 10px;
```
Comment (NotMyFault) [inline war/src/main/less/base/style.less:576]: What exactly is this change for? It adds more space to the last row of all items for no obvious reason to me.

Comment (frankie139506) [reply war/src/main/less/base/style.less:576]: I think it look more tidy, so I add some space in the bottom, if this is redundant I have reverted, please take a look.

File: src/main/scss/base/_style.scss [hunk 2]
```diff
@@ -389,6 +391,24 @@ pre.console {
   p:last-of-type {
```
Comment (timja) [inline src/main/scss/base/_style.scss:391]: I tried removing this as its the cause of the heading spacing issue but I couldn't find a way to make it work whilst keeping no spacing at the bottom of the help file (made worse in help from plugins that have `from plugin ...` as the space was a bit excessive without this.

File: src/main/scss/base/_style.scss [hunk 3]
```diff
@@ -389,6 +391,24 @@ pre.console {
...
     margin-bottom: 0;
   }
+
+  // add spacing above headings except for when its the first element in the help
+  // the need for this is caused by p:last-of-type setting margin-bottom to 0
+  // unfortunately because of the varied markup I wasn't able to find a way to avoid this
+  h1:not(:first-child),
```
Comment (timja) [inline src/main/scss/base/_style.scss:398]: @janfaracik what do you think about this approach and can you think of a simpler one?

File: core/src/main/resources/hudson/model/UsageStatistics/global.groovy [hunk 4]
```diff
@@ -8,7 +8,7 @@ def f=namespace(lib.FormTagLib)
 
 f.section(title: _("Usage Statistics")) {
     if (UsageStatistics.DISABLED) {
-        span(class: "jenkins-not-applicable") {
+        div(class: "jenkins-not-applicable jenkins-!-margin-bottom-2") {
```
Comment (timja) [inline core/src/main/resources/hudson/model/UsageStatistics/global.groovy:11]: Didn't look great as it was touching the line

Before:
image

After:
image

File: core/src/main/resources/hudson/model/UsageStatistics/help-usageStatisticsCollected.jelly [hunk 5]
```diff
@@ -38,24 +38,34 @@
             Each trial has a specific purpose and a defined end date, after which collection stops, independent of the installed versions of Jenkins or plugins.
             Once a trial is complete, the trial results may be aggregated and shared with the developer community.
         </p>
-        <p>
-            The following trials defined on this instance are active now or in the future:
```
Comment (timja) [inline core/src/main/resources/hudson/model/UsageStatistics/help-usageStatisticsCollected.jelly:42]: the help was ending with this as there's no trials currently (and an empty `dl`)

**Review by timja — APPROVED** [review]

**Comment by janfaracik** [comment]
Will take a look 👍

**Comment by timja** [comment]
@janfaracik please take a look

File: core/src/main/resources/hudson/model/UsageStatistics/global.groovy (hunk 4 above)
Comment (janfaracik) [inline core/src/main/resources/hudson/model/UsageStatistics/global.groovy:11]: `jenkins-description` will give the appropriate colour/spacing:

image

Can leave `jenkins-not-applicable` if you need the italic text, don't think its adding anything though.

Comment (timja) [reply core/src/main/resources/hudson/model/UsageStatistics/global.groovy:11]:
```suggestion
        div(class: "jenkins-not-applicable jenkins-description") {
```

File: src/main/scss/base/_style.scss (hunk 3 above)
Comment (janfaracik) [reply src/main/scss/base/_style.scss:398]: Yeah I can't think of anything sadly, could use 'gap' on the parent container but due to how varied the contents are it probably would make things worse :/

**Review by janfaracik — APPROVED** [review]

**Comment by timja** [comment]
/label ready-for-merge

----

This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Pull request merged**